[Kevin Nathanael] iP

kevinmingtarja opened the pull request:
### HAL9000 (Another flavor of Duke)
> "No man ever steps in the same river twice, for it's not the same river and he's not the same man." - Heraclitus

HAL9000 frees your mind by remembering the things you need to do. It's
- text-based
- easy to use
- SUPER FAST

All you need to do is,
1. download it
2. run it
3. add your tasks

And it is **FREE!**

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [x] Managing events

---

You can also use it to practice Java too. Here's the `main` method:

```java
public static void main(String[] args) {
        new Duke("HAL9000").initialize();
}
```

Commit:
Add Gradle support

kevinmingtarja committed:
Finish Level-1, created Chatbot class

kevinmingtarja committed:
Finish Level-2, create Storage class

kevinmingtarja committed:
Level-3, add Task class and renamed some classes

kevinmingtarja committed:
Level-4, created ToDo, Event, and Deadline class

kevinmingtarja committed:
A-TextUiTesting, add semi-automated testing

kevinmingtarja committed:
Level-5, add several Exception classes to handle errors

kevinmingtarja committed:
Level-6, add functionality to delete a Task, along with its error handling

kevinmingtarja committed:
Add save and read from file functionality

kevinmingtarja committed:
Merge pull request #1 from kevinmingtarja/branch-Level-7

Add save and read from file functionality

kevinmingtarja committed:
Refactor Deadline to use LocalDate instead of string

kevinmingtarja committed:
Merge pull request #2 from kevinmingtarja/branch-Level-8

Refactor Deadline to use LocalDate instead of string

kevinmingtarja committed:
Extract Ui class

kevinmingtarja committed:
Extract Storage class

kevinmingtarja committed:
Extract Parser class

kevinmingtarja committed:
Organize into Packages

kevinmingtarja committed:
Add unit tests

kevinmingtarja committed:
Add JavaDoc comments

kevinmingtarja committed:
Merge pull request #3 from kevinmingtarja/branch-A-JavaDoc

Add JavaDoc comments

kevinmingtarja committed:
Tweak code to comply with coding standard

kevinmingtarja committed:
Merge pull request #4 from kevinmingtarja/branch-A-CodingStandard

Tweak code to comply with coding standard

kevinmingtarja committed:
Add find task functionality

kevinmingtarja committed:
Merge pull request #5 from kevinmingtarja/branch-Level-9

Add find task functionality

kevinmingtarja committed:
Merge branch 'add-gradle-support' into branch-A-Gradle

kevinmingtarja committed:
Fix mainClassName in build.gradle

kevinmingtarja committed:
Merge pull request #6 from kevinmingtarja/branch-A-Gradle

Integrate Gradle into the project

kevinmingtarja committed:
Add checkstyle and fix code styles

kevinmingtarja committed:
Merge pull request #7 from kevinmingtarja/branch-A-CheckStyle

Add checkstyle and fix code styles

imposterble commented on src/test/java/duke/DukeTest.java:
```diff
@@ -0,0 +1,12 @@
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Perhaps the position of these 2 statements should be reversed (so that the ordering is consistent with that of other files)?

imposterble commented on src/test/java/duke/DukeTest.java:
```diff
@@ -0,0 +1,12 @@
+
+public class DukeTest {
+    @Test
+    public void dummyTest(){
```
Should there be whitespace before "{"?

imposterble commented on src/test/java/duke/ui/UiTest.java:
```diff
@@ -0,0 +1,39 @@
+    @Test
+    public void print_invalidFormat_exceptionThrown() throws Exception {
+        Ui ui = new Ui();
+        assertThrows(Exception.class, () -> ui.print("%d",  "two"));
```
Should there be only a single space to separate non-whitespace characters (e.g., "two")?

imposterble commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,47 @@
+package duke;
```
I **like** how you packaged the relevant classes into sub-packages

imposterble reviewed (commented):
LGTM! Good work overall and all the best with the remaining iP tasks :smile:

jeremy-7007 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,159 @@
+     */
+    public boolean parseInput(String input) {
+        ui.printDivider();
+        try {
```
Why not try using a switch statement! It might make your code a bit more clean, but otherwise, it is fine as is.

rushilramesh replied:
Agreed, a switch statement would be useful here since there are a significant number of cases.

kdeasymoneysniper replied:
Yup, a switch statement would make this part cleaner, although I would say it is still very readable the way it is now!

jeremy-7007 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,159 @@
+     */
+    private void addItem(String input) throws DukeException {
+        Task newItem = null;
+        if (input.contains("todo")) {
```
Unless I am misunderstanding your code, if you leave it as `input.contains("todo")`, wouldn't there be a bug where the user types, for example, "bla bla todo homework"? The if-condition is fulfilled, but the outcome might not be as expected. Same for other else if statements below.

rushilramesh replied:
Agreed, perhaps it would be better if you only check the first word?

kdeasymoneysniper replied:
Yeah, theres a .startsWith(<string>) method for String classes in Java!

jeremy-7007 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,94 @@
+                boolean isDone = Character.toString(line.charAt(line.indexOf("[", 2) + 1)).equals("X");
+                String name = line.substring(line.indexOf("]", line.indexOf("]") + 1) + 2);
+
+                if (type.equals("T")) {
```
A switch statement can be used here too!

jeremy-7007 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,159 @@
+                throw new DeadlineException();
+            }
+
+            String[] description = parsedInput[1].split("/");
```
If you want to be even more strict with the user inputs, it might be a good idea to check that the "/" is indeed followed with "by" before splitting the input into event name and time. Same for event below.

rushilramesh replied:
Agreed, perhaps it would be better to also specify in your DukeException the format of acceptable user input.

jeremy-7007 reviewed (commented):
Made some suggestions to the Parser class. Otherwise, looks good to me. Good luck for the rest of the iP!

rushilramesh reviewed (commented):
Code makes sense. All the best!

kevinmingtarja committed:
Integrate project with JavaFX

kevinmingtarja committed:
Merge pull request #8 from kevinmingtarja/branch-Level-10

Finish Level 10

kdeasymoneysniper commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,32 @@
+    @Override
+    public void start(Stage stage) {
+        try {
+            FXMLLoader fxmlLoader = new FXMLLoader(Main.class.getResource("/view/MainWindow.fxml"));
```
maybe separate into multiple lines :-)

kdeasymoneysniper reviewed (commented):
Your code is really well thought out and the OOP demonstrated is pretty sick! :-)

kdeasymoneysniper reviewed (commented):
Your code is really well thought out and the OOP demonstrated is pretty sick! :-)

kevinmingtarja committed:
Add assertions

Assertions have not been added yet to the codebase.

Adding assertions allows us to verify the correctness of invariants at
certain parts of our code.

Using assertions are preferable in development as it can catch bugs in
our code.

kevinmingtarja committed:
Improve code quality

Some parts of the code still violates the standard for code quality.

Refactor the code based on pull request reviews from other peers.

kevinmingtarja committed:
Merge pull request #10 from kevinmingtarja/branch-A-Assertions

Add assertions

kevinmingtarja committed:
Merge branch 'master' into branch-A-CodeQuality

kevinmingtarja committed:
Merge pull request #11 from kevinmingtarja/branch-A-CodeQuality

Improve code quality

kevinmingtarja committed:
Add feature to tag tasks

kevinmingtarja committed:
Modularize load storage function

kevinmingtarja committed:
Merge pull request #12 from kevinmingtarja/branch-C-Tagging

Add tagging feature

kevinmingtarja committed:
Improve UI

kevinmingtarja committed:
Fix imports

kevinmingtarja committed:
Fix checkstyle warnings

kevinmingtarja committed:
Change profile picture and add screenshot

kevinmingtarja committed:
Update README

kevinmingtarja committed:
Update README

kevinmingtarja committed:
Set theme jekyll-theme-cayman

kevinmingtarja committed:
Minor fixes

kevinmingtarja committed:
Merge branch 'master' of github.com:kevinmingtarja/ip

kevinmingtarja committed:
Update README

kevinmingtarja committed:
Merge branch 'master' of github.com:kevinmingtarja/ip

kevinmingtarja committed:
Add more error handlings

kevinmingtarja committed:
Fix filename in README

kevinmingtarja committed:
Fix build.gradle
